bpython/simpleeval.py

```diff
@@ -0,0 +1,145 @@
+# encoding: utf-8
+"""simple evaluation of side-effect free code
+
+In order to provide fancy completion, some code can be executed safely.
+
+"""
+
+from ast import *
+from six import string_types
+
+from bpython import line as line_properties
+
+class EvaluationError(Exception):
+    """Raised if an exception occurred in safe_eval."""
+
+
+def safe_eval(expr, namespace):
+    """Not all that safe, just catches some errors"""
+    try:
+        return eval(expr, namespace)
+    except (NameError, AttributeError, SyntaxError):
+        # If debugging safe_eval, raise this!
+        # raise
+        raise EvaluationError
+
+def simple_eval(node_or_string, namespace=None):
+    """
+    Safely evaluate an expression node or a string containing a Python
+    expression.  The string or node provided may only consist of:
+    * the following Python literal structures: strings, numbers, tuples,
+        lists, dicts, booleans, and None.
+    * variable names causing lookups in the passed in namespace or builtins
+    * getitem calls using the [] syntax on objects of the types above
+    * getitem calls on subclasses of the above types if they do not override
+        the __getitem__ method and do not override __getattr__ or __getattribute__
+        (or maybe we'll try to clean those up?)
+
+    The optional namespace dict-like ought not to cause side effects on lookup
+    """
+    if namespace is None:
+        namespace = {}
+    if isinstance(node_or_string, string_types):
+        node_or_string = parse(node_or_string, mode='eval')
+    if isinstance(node_or_string, Expression):
+        node_or_string = node_or_string.body
+    def _convert(node):
+        if isinstance(node, Str):
+            return node.s
+        elif isinstance(node, Num):
+            return node.n
+        elif isinstance(node, Tuple):
+            return tuple(map(_convert, node.elts))
+        elif isinstance(node, List):
+            return list(map(_convert, node.elts))
+        elif isinstance(node, Dict):
+            return dict((_convert(k), _convert(v)) for k, v
+                        in zip(node.keys, node.values))
+        elif isinstance(node, Name):
+            try:
+                return namespace[node.id]
+            except KeyError:
+                return __builtins__[node.id]
+        elif isinstance(node, BinOp) and \
+             isinstance(node.op, (Add, Sub)) and \
+             isinstance(node.right, Num) and \
+             isinstance(node.right.n, complex) and \
+             isinstance(node.left, Num) and \
+             isinstance(node.left.n, (int, long, float)):
```

Comment:
Hm, working in Python 2 but not 3, so looks like you've hit exactly on the problem. I'll look at how much code can be shared between the two, they might have to be separate implementations.